Ensure third party frames do not have cookie access from Cookie Store API change listener

RupinMittal opened the pull request:
#### 0df5a7c46418c0afa242a7c4d07761a813251e23
<pre>
Ensure third party frames do not have cookie access from Cookie Store API change listener
<a href="https://bugs.webkit.org/show_bug.cgi?id=283985">https://bugs.webkit.org/show_bug.cgi?id=283985</a>
<a href="https://rdar.apple.com/140446608">rdar://140446608</a>

Reviewed by Sihui Liu.

Previously, a third party iframe could register a cookie-change event listener.
Then, if another page set a cookie, the iframe would get a notification about
the cookie change. But third party iframes should not have this cookie access.

The fix: when the iframe tries to register a change listener, a listener won&apos;t
be registered if the iframe&apos;s cookie access should be blocked.

Furthermore, we add allowsFirstPartyForCookies() and shouldBlockCookies() checks
in the Network Process because the Web Process may not have all the information
to know if it should block cookie access. Since it&apos;s possible that the checks
in the Web Process pass and those in the Network Process fail, we refactor the
code so that the Network Process returns a boolean indicating if the checks passed.
So the Web Process will only add the listener to it&apos;s local map of listeners if
the checks in the Network Process pass and the listener is indeed registered.

Since WebCookieCache&apos;s call to add a cookie change listener happens only if it
passes the Web Process&apos; shouldBlockCookies check in WebCookieJar, we don&apos;t need
to do the Web Process check again for it. So we add a new addChangeListenerWithAccess
function that doesn&apos;t do the Web Process check but does still do the Network
Process checks.

* LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener-expected.txt: Added.
* LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener.html: Added.
* LayoutTests/http/wpt/cookie-store/resources/third_party_iframe_change_listener.html: Added.
* LayoutTests/http/wpt/cookie-store/resources/third_party_iframe_change_listener_helper.html: Added.
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::addChangeListener):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::startListeningForCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::cookiesForDOM):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::addChangeListenerWithAccess):
(WebKit::WebCookieJar::addChangeListener):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/287462@main">https://commits.webkit.org/287462@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48bd52c94ec7988c79ce3222c7a9275be953439e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33106 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84247 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30730 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62302 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20146 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82784 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52352 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72595 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42610 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26747 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29168 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70831 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27212 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85670 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6953 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4847 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70558 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7123 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68441 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69798 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/17413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12724 "Passed tests") | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12328 "Failed to push commit to Webkit repository") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6907 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12527 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6776 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8579 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_third_party_iframe_change_listener.html:
```diff
@@ -0,0 +1,28 @@
+<meta charset='utf-8'>
+<script src='/resources/testharness.js'></script>
+<script src='/resources/testharnessreport.js'></script>
+<iframe id='iframe' src='http://127.0.0.1:8800/cookie-store/resources/third_party_iframe_change_listener.html'></iframe>
```
Since his is a WPT test, you shouldn't be hard-coding host names / ports, that's not OK outside WebKit-only tests.

You'll have to use get-host-info.sub.js and `get_host_info()`. You can see examples in other WPT tests.

cdumez replied:
But also, cookie partitioning (or blocking) in third-party iframes is not consistent across browsers so it may just be safer to move this test to http/tests/wpt.

nt1m replied:
Could be worth making a .tentative.html test and exporting to WPT

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -156,7 +156,7 @@ String WebCookieJar::cookies(WebCore::Document& document, const URL& url) const
 
     if (isEligibleForCache(*webFrame, document.firstPartyForCookies(), url))
-        return m_cache.cookiesForDOM(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies);
+        return m_cache.cookiesForDOM(document, document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies);
```
Maybe we can omit the second parameter and let the implementation call `document.firstPartyForCookies()` now that we pass it the `document`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -404,8 +404,16 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+        return;
+
+    ApplyTrackingPrevention applyTrackingPreventionInNetworkProcess = ApplyTrackingPrevention::No;
+    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), document.cookieURL(), applyTrackingPreventionInNetworkProcess))
```
@szewai probably remembers the details but I think we need a similar check on the network process side.

See for example this code running in the network process when `document.cookie` is called:
```
RetainPtr<NSArray> NetworkStorageSession::cookiesForURL(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) const
{
    if (applyTrackingPrevention == ApplyTrackingPrevention::Yes && shouldBlockCookies(firstParty, url, frameID, pageID, shouldRelaxThirdPartyCookieBlocking))
        return nil;
    return httpCookiesForURL(cookieStorage().get(), firstParty, sameSiteInfo, url);
}
```

It checks shouldBlockCookies() too.

My bet is that we need a similar check on network process side when registering a cookie listener.

geoffreygaren replied:
Since a compromised WebContent process can skip all these checks, a check in the Network process is essential.

Probably the check belongs inside the handler for SubscribeToCookieChangeNotifications.

cdumez replied:
To be clear, I'm not encouraging to move the check to the network process but rather add an additional one in the NetworkProcess, just like we do for the `document.cookie` code path. I think we should be consistent.

RupinMittal replied:
@szewai I've added the Network Process checks to the `addChangeListener` code path. Do we also want them in the `removeChangeListener` path? I thought about doing this, but it didn't seem feasible in some places. For example, where would `WebCookieCache::clear()` or `NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess()` get the needed document?

Also, if the WebProcess check passes, but the Network Process check fails, we would have already added the listener to `WebCookieaJar::m_changeListeners`. Is this ok or does the Network Process need to IPC back to let WebProcess know if it's ok to add the listener to the map? The same issue exists with the two checks in the Network Process. If the check in NetworkConnectionToWebProcess passes but the check the NetworkStorageSession fails, we've already added the listener to `NetworkConnectionToWebProcess ::m_hostsWithCookieListeners`. Is this ok?

szewai replied:
> Do we also want them in the removeChangeListener path?

We don't need to perform check for `removeChangeListener` since it does not give web process more access it does not already have.

> Also, if the WebProcess check passes, but the Network Process check fails, we would have already added the listener to WebCookieaJar::m_changeListeners. Is this ok or does the Network Process need to IPC back to let WebProcess know if it's ok to add the listener to the map?

With existing implementation, technically web process needs to know whether the subscription succeeds or not, because it sends only one message for one host. (i.e. if the subscription fails in network process but web process does not know, web process will not try to ask for the subscription again.)

> The same issue exists with the two checks in the Network Process. If the check in NetworkConnectionToWebProcess passes but the check the NetworkStorageSession fails, we've already added the listener to NetworkConnectionToWebProcess ::m_hostsWithCookieListeners. Is this ok?

`NetworkConnectionToWebProcess::m_hostsWithCookieListeners` looks unnecessary to me, since `NetworkStorageSession` already tracks the information. Maybe we can just drop it.

charliewolfe replied:
> We don't need to perform check for removeChangeListener since it does not give web process more access it does not already have.

Having equivalent checks for `removeChangeListener` would prevent a web process from removing cookie listeners it does not own. No idea why that’s something a compromised web process would want to do, so maybe it's not important.

szewai replied:
Cookie change observer in network process is stored by connection, so a WP cannot change the subscription state of another WP.

RupinMittal replied:
I've added the functionality where the Network Process returns a boolean telling the Web Process if the checks passed and listener was indeed added or not. 

It seems that `NetworkConnectionToWebProcess::m_hostsWithCookieListeners` is passed to `NetworkStorageSession::stopListeningForCookieChangeNotifications` to stop the listening. This seems important,  so I didn't remove that map yet. I've updated the code to only add the host to the map if `shouldBlockCookies` in `NetworkStorageSession` is not true. What do we think of that?

szewai replied:
> This seems important, so I didn't remove that map yet. I've updated the code to only add the host to the map if shouldBlockCookies in NetworkStorageSession is not true. What do we think of that?

This is not important as `NetworkStorageSession` keeps a map of { host, `NetworkConnectionToWebProcess`s}, and `NetworkStorageSession::stopListeningForCookieChangeNotifications(*this, m_hostsWithCookieListeners)` basically asks `NetworkStorageSession` to clear `NetworkConnectionToWebProcess` entries from the map, so hosts are not required. 
Tracking hosts might be slightly performant since we don't need to iterate all entries of map; but that should not be necessary since it only happens in `~NetworkConnectionToWebProcess()`.

RupinMittal replied:
Ok that makes sense. Updated!

rr-codes commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
```diff
@@ -65,14 +65,14 @@ static String cookiesToString(const Vector<WebCore::Cookie>& cookies)
 }
 
-String WebCookieCache::cookiesForDOM(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, FrameIdentifier frameID, PageIdentifier pageID, IncludeSecureCookies includeSecureCookies)
+String WebCookieCache::cookiesForDOM(const WebCore::Document& document, const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, FrameIdentifier frameID, PageIdentifier pageID, IncludeSecureCookies includeSecureCookies)
```
the WebCore:: isn't needed since this is already within the namespace

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
```diff
@@ -65,14 +65,14 @@ static String cookiesToString(const Vector<WebCore::Cookie>& cookies)
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
         if (!hasCacheForHost)
-            WebProcess::singleton().protectedCookieJar()->addChangeListener(host, *this);
+            WebProcess::singleton().protectedCookieJar()->addChangeListener(document, host, *this);
```
Cookie access should already be checked when `WebCookieCache` function is invoked, so we don't need to do it again. We can add another `addChangeListenerWithAccess` (which is the same as the current `addChangeListener`) and use it here.

RupinMittal replied:
That's a good point. I was going to do this and then realized that to add the Network Process checks, we need to pass in the document anyways. So I didn't create another function. Based on the new parameters being passed in, do you think we still want another `addChangeListenerWithAccess` function that doesn't do the Web Process checks?

szewai replied:
We could have a `addChangeListenerWithAccess` function that takes network process needs (and you can get them from `WebCookieCache::cookiesForDOM` arguments).

RupinMittal replied:
Added!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_third_party_iframe_change_listener.html:
```diff
@@ -0,0 +1,28 @@
+    const iframe = document.getElementById('iframe');
+
+    await new Promise(resolve => {
+        iframe.addEventListener("load", resolve);
```
```suggestion
        iframe.addEventListener("load", resolve, { once: true });
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_third_party_iframe_change_listener.html:
```diff
@@ -0,0 +1,28 @@
+        window.addEventListener('message', event => {
+            assert_equals(event.data, false);
+            resolve();
+        });
```
```suggestion
        }, { once: true });
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/resources/third_party_iframe_change_listener.html:
```diff
@@ -0,0 +1,13 @@
+
+        window.addEventListener('message', event => {
+            parent.postMessage(changeEventFired, '*');
+        });
```
```suggestion
        }, { once: true });
```

RupinMittal replied:
Added!

nt1m reviewed (commented):
I'd add `{ once: true }` in places where we expect only one event

cdumez commented on LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener.html:
```diff
@@ -0,0 +1,40 @@
+}
+
+promise_test(async t => {
+    const iframe = await with_iframe(notSameSiteBaseURL + "/resources/third_party_iframe_change_listener.html");
```
The top document is not HTTPS but the iframe is HTTPS? Does this even work? I'm a little surprised in a non-https document can load https subframes..

RupinMittal replied:
Huh the test was working but that's s good point. Just updated it to use HTTP!

nt1m commented on LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener.html:
@cdumez Should we consider exporting this to WPT as a tentative test?

cdumez replied:
I don't know what the current state of third-party cookies handling in specs. It seems risky. This is an escape fix, let's not delay this PR because of this. We can follow-up and upstream if needed but we should discuss with @annevk first.

annevk replied:
If you end the test file name with `.tentative.html` you can pretty much land whatever.

And I think it would be useful to have coverage for this.

(We're still trying to figure out exactly how to deal with third-party cookie deprecation, but I think what's likely is that we'll assume they're blocked by default across all browsers and some WebDriver API is needed to change from that default for tests that want to test third-party cookies in some way.)

RupinMittal commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -52,7 +52,7 @@ messages -> NetworkConnectionToWebProcess WantsDispatchMessage {
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
-    SubscribeToCookieChangeNotifications(String host) AllowedWhenWaitingForSyncReply
+    SubscribeToCookieChangeNotifications(URL url, URL firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, enum:bool WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) -> (bool result) Synchronous AllowedWhenWaitingForSyncReplyDuringUnboundedIPC
```
Is using `AllowedWhenWaitingForSyncReplyDuringUnboundedIPC` correct? The name sounds close to `AllowedWhenWaitingForSyncReply` which we were using before

RupinMittal replied:
We went back to async IPC and so we can continue to use `AllowedWhenWaitingForSyncReply`.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -902,13 +902,20 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
-void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const String& host)
+void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const URL& url, const URL& firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, CompletionHandler<void(bool)>&& completionHandler)
 {
+    MESSAGE_CHECK(protectedNetworkProcess()->allowsFirstPartyForCookies(m_webProcessIdentifier, firstParty));
```
```suggestion
    MESSAGE_CHECK_COMPLETION(protectedNetworkProcess()->allowsFirstPartyForCookies(m_webProcessIdentifier, firstParty), completionHandler(false));
```

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -404,14 +404,38 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+        auto listenersForHost = m_changeListeners.get(host);
+        if (listenersForHost.computeSize() > 1 || listenersForHost.contains(listener))
+            return false;
+    }
```
Let's use `find()` to reduce one lookup.
`if (auto iter = m_changeListeners.find(host); iter != m_changeListeners.end())`

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -52,7 +52,7 @@ messages -> NetworkConnectionToWebProcess WantsDispatchMessage {
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
-    SubscribeToCookieChangeNotifications(String host) AllowedWhenWaitingForSyncReply
+    SubscribeToCookieChangeNotifications(URL url, URL firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, enum:bool WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) -> (bool result) Synchronous AllowedWhenWaitingForSyncReplyDuringUnboundedIPC
```
This needs to be synchronous?

charliewolfe replied:
Seems like this should not be made synchronous especially if it's also used for cookie caching.

RupinMittal replied:
Reverted back to async. We now use sendWithAsyncReply.

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
```diff
@@ -65,14 +65,14 @@ static String cookiesToString(const Vector<WebCore::Cookie>& cookies)
 }
 
-String WebCookieCache::cookiesForDOM(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, FrameIdentifier frameID, PageIdentifier pageID, IncludeSecureCookies includeSecureCookies)
+String WebCookieCache::cookiesForDOM(const Document& document, const SameSiteInfo& sameSiteInfo, const URL& url, FrameIdentifier frameID, PageIdentifier pageID, IncludeSecureCookies includeSecureCookies)
```
Why is the parameter changed? `addChangeListenerWithAccess` only need `firstPartyURL` and it's provided in original function.

RupinMittal replied:
Good point--fixed!

szewai commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -902,13 +902,20 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
-    m_hostsWithCookieListeners.add(host);
 
+    bool startedListening = false;
     if (auto* networkStorageSession = storageSession())
```
`CheckedPtr`

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -741,26 +745,20 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
         m_subscribedDomainsForCookieChanges = adoptNS([[NSMutableSet alloc] init]);
     else if ([m_subscribedDomainsForCookieChanges containsObject:(NSString *)host])
-        return;
+        return false;
```
This should return true.
`m_subscribedDomainsForCookieChanges` is shared by all `CookieChangeObserver` (i.e. all `NetworkConnectionToWebProcess` / web processes). If we return false here, only the first observer can subscribe.

RupinMittal replied:
Oh right, that's a good point. Fixed!

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -404,14 +404,41 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
+
+    if (auto iter = m_changeListeners.find(host); iter != m_changeListeners.end()) {
+        auto listenersForHost = iter->value;
+        if (listenersForHost.computeSize() > 1 || listenersForHost.contains(listener))
```
We probably need to drop `listenersForHost.computeSize() > 1` and do the access check in network process for each listener. 
`listenersForHost.computeSize() > 1` only means the web process has a listener which successfully registers subscription for the host, but the new listener's access does not necessarily match the existing listener's access (unless the two listeners have the same url, firstPartyURL, frameID, pageID and shouldRelaxThirdPartyCookieBlocking).

RupinMittal replied:
Makes sense--fixed!

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -902,22 +902,21 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
-
     if (auto* networkStorageSession = storageSession())
-        networkStorageSession->stopListeningForCookieChangeNotifications(*this, HashSet<String> { host });
+        networkStorageSession->stopListeningForCookieChangeNotifications(*this);
```
Why is this change OK? `stopListeningForCookieChangeNotifications()` will unregistered ALL listeners now, not just the one for the given host. The host is now ignored in the new code which doesn't seem to make sense.

szewai replied:
This change is not ok. I asked for removing `m_hostsWithCookieListeners` because `NetworkStorageSession` already keeps a map of { host, `NetworkConnectionToWebProcess`s}, and it's unnecessary to keep information in two places (`NetworkStorageSession` and `NetworkConnectionToWebProcess`).
i.e. `networkStorageSession->stopListeningForCookieChangeNotifications(*this, m_hostsWithCookieListeners);` can be done without `m_hostsWithCookieListeners` because it just means to remove all entries for the `NetworkConnectionToWebProcess` to `NetworkStorageSession`.

But I didn't ask for removing `host` from `stopListeningForCookieChangeNotifications`. `networkStorageSession->stopListeningForCookieChangeNotifications(*this, HashSet<String> { host });` is about to remove one entry to `NetworkConnectionToWebProcess `, and `NetworkStorageSession ` need the host information. 

I am ok with not removing `m_hostsWithCookieListeners` in this patch if it means too much work. We could do it in followup with test to verify we don't remove subscription accidentally. (Note if we keep `m_hostsWithCookieListeners`, we only want to update it when `startedListening` is true).

RupinMittal replied:
That's my mistake--the change to remove `hosts` from `stopListeningForCookieChangeNotifications` was wrong. I've reverted that and also brought back `m_hostsWithCookieListeners` since we think removing it is too big of a refactoring for this patch. Since we're keeping `m_hostsWithCookieListeners`, I made sure we only update it when `startedListening` is true. Let me know how this looks now.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -489,9 +489,6 @@ class NetworkConnectionToWebProcess final
     NetworkMDNSRegister m_mdnsRegister;
 #endif
-#if HAVE(COOKIE_CHANGE_LISTENER_API)
-    HashSet<String> m_hostsWithCookieListeners;
```
This PR is making too many changes. It is as likely to introduce new bugs as fixing the particular bug we care about. It's been days, please try and make a more targeting fix we can land today.

RupinMittal replied:
Yeah, that's fair. I've just updated it with less refactoring (keeping `m_hostsWithCookieListeners`).

cdumez reviewed (changes requested):
Too many changes, too risky for a critical bug fix.

szewai commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -902,13 +902,20 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
+    MESSAGE_CHECK_COMPLETION(protectedNetworkProcess()->allowsFirstPartyForCookies(m_webProcessIdentifier, firstParty), completionHandler(false));
+
+    auto host = url.host().toString();
     ASSERT(!m_hostsWithCookieListeners.contains(host));
```
My understanding is you need to drop this assertion because web process could send request for the same host, but with different url / firstParty / frameID / pageID / shouldRelaxThirdPartyCookieBlocking with latest impl

RupinMittal replied:
That makes sense. Fixed!

szewai reviewed (approved):
r=me if bots are happy

RupinMittal commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -902,13 +902,17 @@ void NetworkConnectionToWebProcess::domCookiesForHost(const URL& url, Completion
-void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const String& host)
+void NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications(const URL& url, const URL& firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, CompletionHandler<void(bool)>&& completionHandler)
 {
-    ASSERT(!m_hostsWithCookieListeners.contains(host));
```
As Sihui explained, we dropped this assertion because now the web process can send request for the same host, but with different url / firstParty / frameID / pageID / shouldRelaxThirdPartyCookieBlocking. 

This means that in `NetworkStorageSessionCocoa::startListeningForCookieChangeNotifications`, we may attempt to add this same observer again. Prior this code change, we had an assert to ensure this never happened. Which was ok, because if the same host was encountered again, it would first fail here in `NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications`. Without this first assert, some tests fail since they try to subscribe the same observer with different url, etc.

But since we still want to try listening for cookies with the potential different url and so on, we no longer want to have the assert for the observer in `startListeningForCookieChangeNotifications`, so we fix the failures by removing the assert but still only adding the observer if it hasn't already been added.

RupinMittal reviewed (commented):
Made a fix for the layout test failures, let me know how this looks.

RupinMittal commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -728,23 +728,29 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
-    observers.add(observer);
+
+    if (!observers.contains(observer))
+        observers.add(observer);
```
The comment above is about this change. Let me know if this is correct.

szewai replied:
You don't need `if (!observers.contains(observer))`. `observers.add(observer);` will only add item if it doesn't exist.

RupinMittal replied:
Fixed!

RupinMittal committed:
Ensure third party frames do not have cookie access from Cookie Store API change listener
https://bugs.webkit.org/show_bug.cgi?id=283985
rdar://140446608

Reviewed by Sihui Liu.

Previously, a third party iframe could register a cookie-change event listener.
Then, if another page set a cookie, the iframe would get a notification about
the cookie change. But third party iframes should not have this cookie access.

The fix: when the iframe tries to register a change listener, a listener won't
be registered if the iframe's cookie access should be blocked.

Furthermore, we add allowsFirstPartyForCookies() and shouldBlockCookies() checks
in the Network Process because the Web Process may not have all the information
to know if it should block cookie access. Since it's possible that the checks
in the Web Process pass and those in the Network Process fail, we refactor the
code so that the Network Process returns a boolean indicating if the checks passed.
So the Web Process will only add the listener to it's local map of listeners if
the checks in the Network Process pass and the listener is indeed registered.

Since WebCookieCache's call to add a cookie change listener happens only if it
passes the Web Process' shouldBlockCookies check in WebCookieJar, we don't need
to do the Web Process check again for it. So we add a new addChangeListenerWithAccess
function that doesn't do the Web Process check but does still do the Network
Process checks.

* LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener-expected.txt: Added.
* LayoutTests/http/wpt/cookie-store/cookieStore_third_party_iframe_change_listener.html: Added.
* LayoutTests/http/wpt/cookie-store/resources/third_party_iframe_change_listener.html: Added.
* LayoutTests/http/wpt/cookie-store/resources/third_party_iframe_change_listener_helper.html: Added.
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::addChangeListener):
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::startListeningForCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::subscribeToCookieChangeNotifications):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::cookiesForDOM):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::addChangeListenerWithAccess):
(WebKit::WebCookieJar::addChangeListener):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: https://commits.webkit.org/287462@main
